content/posts/integration-test-code-coverage-in-go/index.md

```diff
@@ -0,0 +1,215 @@
+	go killServer.Start()  
+	go runService(ctx)
+
+	<-ctx.Done()
+
+	killServer.server.Shutdown(context.Background())
+}
+```
+
+Next, we can compile the service with: `go test -c ./ -cover -covermode=count -coverpkg=./...` to create a `.test` binary, and run it with:
+
+```bash
+./my-service.test -test.coverprofile my-service.out
+# server started...
+```
+
+Now our server is running as it normally would, but with go's test tooling built in. All of the code it runs is counted for code coverage now.
+
+In another terminal, we run some tests and then kill it with the HTTP signal.
+
+```bash
+./run-tests.sh
+
+curl localhost:19999 # kills the server in the other terminal
+```
+
+A file named `my-service.out` should be created with coverage information from our run.
+
+
+Using `go test -c` to build test binaries is a nice little trick I learned about recently. I suppose we could accomplish the same thing by just running all our servers with the `go test` command directly, instead of compiling them into binary, but then we would need go installed in our containers.
+
+## Part 2: Running Things
+
+Now that our services are built in test mode, we can add them to our Dockerfile, with a little wrapper script to restart them endleslessly.
+
+```Dockerfile
+FROM debian
+
+WORKDIR /
+# coverage test
+COPY ./server.test /server.test
+COPY ./server_with_restart.sh /server_with_restart.sh
+
+ENTRYPOINT [ "/server_with_restart.sh" ]
+```
+
+server_with_restart.sh
+
+```sh
+#!/bin/sh
+
+while true; do
+    echo "Test server started. Built with 'go test -cover -c'"
+
+    /server.test \
+        -test.coverprofile=cover.out \
+        "$@" || exit 1; # exit if process exited with error
+    
+    # Server was probably stopped by http kill signal
+    # to collect code coverage. Restart.        
+    echo "Server restarting.."
+done
+```
+
+Now start those containers, and run your tests.
+
+
+## Part 3: Collecting Coverage
+
+Now to generate, download and merge our coverage.
+
+To trigger generation, we just have to send an HTTP request to our killserver on port 19999. Not too hard, we can just install curl in our kubernetes container, and run it from the outside with exec.
+
+```bash
+kubectl exec my-service -- curl -s localhost:19999
+```
+
+Then copy the file out of the container:
+
+```bash
+kubectl exec -i my-service -- cat /cover.out > cover/my-service.out
+```
+
+Do this for each service, and then merge the coverage profiles together.
+
+```bash
+go install github.com/wadey/gocovmerge@latest
+
+gocovmerge cover/*.out > cover/merged.cov
+
+# Output Total Coverage 
+go tool cover -func=cover/merged.cov | grep -E '^total\:' | sed -E 's/\s+/ /g'
+```
+
+Note: We use a `.cov` extention on the `merged.cov` to make it easy to use `cover/*.out` in scripting.
+
+And that's about it. For us at Manabie, we are using this final coverage percentage in our Github Actions pipeline to ensure that our integration test coverage does not drop. 
```

Comment:
"does not drop" I think not a term here, the point is make sure we didn't forget to test something.
I know they're similar but I feel explain the end goal is better.